File: Lab-6/lab6.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Course: Understanding Deep Neural Networks
% Teacher: Zhang Yi
% Student: 陈逸韬
% ID: 2020141460308
%
% Lab 6 - Sequence auto-complete
%
% Task:
% Design a multi-target outputs neural network to learn to complete sequence.
% The first two items of a sequence uniquely determine the remaining four.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% clear workspace and close plot windows
clear;
close all;
% prepare the data set
train_seq = {
    'AA1212' 'AC1231' 'AD1221' 'AE1213'
    'BA2312' 'BB2323' 'BC2331' 'BE2313'
    'CB3123' 'CC3131' 'CD3121' 'CE3113'
    'DA2112' 'DB2123' 'DC2131' 'DD2121'
    'EA1312' 'EB1323' 'ED1321' 'EE1313'
    };
test_seq = {
    'AB1223' 'BD2321' 'CA3112' 'DE2113' 'EC1331'
    };
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Your code BELOW
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% encode datasets
%%
%%
% prepare training data
% prepare testing data
% choose parameters
max_iter=100;
lr=0.1;
f = @(s) 1 ./ (1 + exp(-s)); 
df = @(s) f(s) .* (1 - f(s)); 
% define the network architecture
N = [5,5,3,3,3,3];
% initialize weights
%%
W = cell(1,4);
for i=1:4
    W{i}=0.1*randn(3,13);
end
A=cell(1,4);
Z=cell(1,4);
X=cell(1,5);
Y=cell(1,4);
delta=cell(1,4);
dw=cell(1,4);
%%
% train
J = [];
Acc = [];
%%
% loop until converge
for iter = 1:max_iter
    % for each mini-batch
    [m,n]=size(train_seq);
    for i = 1:m
        for j=1:n
            x=train_seq{i,j};
            [encode_x,encode_idx]=encode(x);
            % forward computation
            AB=encode_x{1};
            X{1}=[zeros(3,1) ; AB ];
            for ii = 1:4
               Y{ii}=encode_x{ii+1}; 
            end
            for ii = 1:4
                [A{ii} ,Z{ii}]=fc(W{ii},X{ii});
                X{ii+1}=[A{ii};AB];
            end
            % backward computation (need some attention here)
            delta{4}=(A{4}-Y{4}).*df(Z{4});
            dw{4}=delta{4} * X{4}';
            for ii =3:-1:1
               delta{ii}=[df(Z{ii}) ; ones(10,1)] .* W{ii+1}' * delta{ii+1} + [ df(Z{ii}).*(A{ii}-Y{ii}) ; zeros(10,1)];
               delta{ii}=delta{ii}(1:3);
               dw{ii}=delta{ii} * X{ii}';
            end
            
            for ii =1:4
                W{ii}=W{ii}-lr*dw{ii};
            end
            % update weight
            
            % cost function on train batch (sums from all layers)
            loss=0;
            for ii = 1:4
                loss=loss+(1/2)*sum((A{ii}-Y{ii}).^2);
            end
            J = [J loss];
            plot(J,'-b');drawnow
            % accuary on train batch
%             batch_Acc = [];
%             Acc = [Acc batch_Acc];
        end
    end
    % optionally you can display J and Acc on-the-fly
end
plot(1:2000,J,'-b');
save model
save model.mat W n Acc J
%%
% test
% test on training set
total_num=20;num=0;
[m,n]=size(train_seq);
for i = 1:m
    for j=1:n
        x=train_seq{i,j};
        [encode_x,encode_index]=encode(x);
        label=encode_index(3:6);
        AB=encode_x{1};
        X{1}=[zeros(3,1) ; AB ];
        for ii = 1:4
            [A{ii} ,Z{ii}]=fc(W{ii},X{ii});
            X{ii+1}=[A{ii};AB];
        end
        Y_pre=int32(zeros(1,4));
        for ii = 1:4
           [data,index]=max(A{ii});
           Y{ii}=index; 
           Y_pre(ii)=index;
        end
        
         if isequal(label,Y_pre)
             num=num+1;
         end
    end
end
%%
train_acc=num/total_num;
% test on testing set
test_acc =0;
test_num=0
for i = 1:5
    x=test_seq{i};
    [encode_x,encode_index]=encode(x);
    label=encode_index(3:6);
    AB=encode_x{1};
    X{1}=[zeros(3,1) ; AB ];
    for ii = 1:4
        [A{ii} ,Z{ii}]=fc(W{ii},X{ii});
        X{ii+1}=[A{ii};AB];
    end
    Y_pre=int32(zeros(1,4));
    for ii = 1:4
       [data,index]=max(A{ii});
       Y{ii}=index; 
       Y_pre(ii)=index;
    end

     if isequal(label,Y_pre)
         test_num=test_num+1;
     end
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Your code ABOVE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
test_acc=test_num/5;
% display results
fprintf('Accuracy on training dataset is %f%%\n', train_acc*100);
fprintf('Accuracy on testing dataset is %f%%\n', test_acc*100);
%%
B=[];
for i = 1:4
    B=[ B A{i}];
end
%%
disp(label);
disp(B);

File: Lab-6/lab6.py
import matplotlib.pyplot as plt
import numpy as np
import math, pickle
from units import encode_datalist, prepare_data, fc, bc, cost, accuracy

# prepare the data set
train_list = [
    'AA1212', 'AC1231', 'AD1221', 'AE1213',
    'BA2312', 'BB2323', 'BC2331', 'BE2313',
    'CB3123', 'CC3131', 'CD3121', 'CE3113',
    'DA2112', 'DB2123', 'DC2131', 'DD2121',
    'EA1312', 'EB1323', 'ED1321', 'EE1313'
]
test_list = [
    'AB1223', 'BD2321', 'CA3112', 'DE2113', 'EC1331'
]

# encode datasets
train_list_encoded = encode_datalist(train_list)
test_list_encoded = encode_datalist(test_list)

train_size = 20
# prepare training data
trainData, trainLabels = prepare_data(train_list)
print(trainData.shape, trainLabels.shape)

# prepare testing data
testData, testLabels = prepare_data(test_list)
print(testData.shape, testLabels.shape)

# choose parameters
L = 5
# define the network architecture
layer_size = [10, 6, 6, 6, 6]

# initialize weights
w = {}
for l in range(1, L):
    w[l] = 2 * np.random.randn(layer_size[l] - 3, layer_size[l-1])
alpha = 0.1  # initialize learning rate

# train
J = []
Acc = []
max_epoch = 500 # number of training epoch 20
mini_batch = 4 # number of sample of each mini batch 4
# loop until converge
for i in range(max_epoch):
    idxs = np.random.permutation(train_size)
    # for each mini-batch
    for k in range(math.ceil(train_size / mini_batch)):
        start_idx = k*mini_batch 
        end_idx = min((k+1)*mini_batch, train_size)
        
        a, z, delta = {}, {}, {}
        batch_indices = idxs[start_idx:end_idx]
        a[1] = trainData[:, batch_indices]
        y = trainLabels[:, batch_indices]
        # forward computation
        for l in range(1, L):
            y_out = (l - 1) * 3
            a[l+1], z[l+1] = fc(w[l], a[l], y[y_out - 3:y_out])

        delta[L] = (a[L][-3:] - y[-3:]) * (a[L]*(1-a[L]))[-3:]
        # backward computation (need some attention here)
        for l in range(L-1, 1, -1):
            y_out = (l - 1) * 3
            delta[l] = bc(w[l], z[l], delta[l+1], a[l][-3:] - y[y_out - 3:y_out])
        
        # update weight
        for l in range(1, L):
            #delat_sup_list = [0] * 4
            #delta_sup = np.array(delat_sup_list).reshape(1, 4)
            #delta_sup = np.concatenate((delta_sup, delta[l+1]), axis=0)

            y_out = (l - 1) * 3
            a_l_sup = np.concatenate((y[y_out - 3:y_out], a[l]), axis=0)

            grad_w = np.dot(delta[l+1], a_l_sup.T)
            w[l] = w[l] - alpha*grad_w

        # cost function on train batch (sums from all layers)
        batch_J = cost(a, y)/mini_batch
        J.append(batch_J)

        # accuary on train batch
        batch_Acc = accuracy(a, y)
        Acc.append(batch_Acc)
    
    # optionally you can display J and Acc on-the-fly
    # plot(J,'-b')
    # drawnow
    if i % 50 == 0:
        a[1] = testData
        y = testLabels
        # forward computation 
        for l in range(1, L):
            y_out = (l - 1) * 3
            a[l+1], z[l+1] = fc(w[l], a[l], y[y_out - 3:y_out])
        print(i, "training acc:", Acc[-1], 'test acc:', accuracy(a, y))


# save model
plt.figure()
plt.plot(J)
plt.savefig("J.png")
plt.close()
plt.figure()               
plt.plot(Acc)
plt.savefig("Acc.png")
plt.close()
# Step 8: Store the Network Parameters
# save model
model_name = 'model.pkl'
with open(model_name, 'wb') as f:
    pickle.dump([w, layer_size], f)
print("model saved to {}".format(model_name))
